cirq/experiments/qubit_characterizations.py

```diff
@@ -0,0 +1,567 @@
+    rho_01_im = numpy.mean(results.measurements['z']) - 0.5
+
+    circuit_y = circuit.copy()
+    circuit_y.append(ops.Y(qubit) ** -0.5)
+    circuit_y.append(ops.measure(qubit, key='z'))
+    results = sampler.run(circuit_y, repetitions=num_shots)
+    rho_01_re = 0.5 - numpy.mean(results.measurements['z'])
+
+    rho_01 = rho_01_re + 1j * rho_01_im
+    rho_10 = numpy.conj(rho_01)
+
+    rho = numpy.array([[rho_00, rho_01], [rho_10, rho_11]])
+
+    return StateTomographyResult(rho)
+
+
+def two_qubit_state_tomography(sampler: sim.SimulatesSamples,
+                               q_0: devices.GridQubit, q_1: devices.GridQubit,
+                               circuit: circuits.Circuit, num_shots: int) \
+        -> StateTomographyResult:
+    """
+    Two-qubit state tomography.
+
+    The density matrix of the output state of a two-qubit circuit is
+    determined by z-basis measurements preceded by nine different
+    combinations of I, X/2 and Y/2 operations on the two qubits. The sixteen
+    elements of the density matrix are obtained from the least-square
+    solution to a system of linear equations that may be constructed from the
+    measurement results.
+
+    See Vandersypen and Chuang, Rev. Mod. Phys. 76, 1037 for details.
+
+    Args:
+        sampler: The quantum engine or simulator to run the circuits.
+        q_0: The first qubit under test.
+        q_1: The second qubit under test.
+        circuit: The circuit to execute on the qubits before tomography.
+        num_shots: The number of measurements for each basis rotation.
+
+    Returns:
+        A StateTomographyResult object that stores and plots the density matrix.
+    """
+
+    def _measurement(circ):
+        circ.append(ops.measure(q_0, q_1, key='z'))
+        results = sampler.run(circ, repetitions=num_shots)
+        bit_strings = [r for r in results.measurements['z']]
+        p_00 = 0.0
+        p_01 = 0.0
+        p_10 = 0.0
+        for bits in bit_strings:
+            if not bits[0] and not bits[1]:
+                p_00 += 1.0 / num_shots
+            elif not bits[0] and bits[1]:
+                p_01 += 1.0 / num_shots
+            elif bits[0] and not bits[1]:
+                p_10 += 1.0 / num_shots
+        return [p_00, p_01, p_10]
+
+    sigma_0 = numpy.eye(2) / 2.0
+    sigma_1 = numpy.array([[0.0, 1.0], [1.0, 0.0]]) / 2.0
+    sigma_2 = numpy.array([[0.0, -1.0j], [1.0j, 0.0]]) / 2.0
+    sigma_3 = numpy.array([[1.0, 0.0], [0.0, -1.0]]) / 2.0
+    sigmas = [sigma_0, sigma_1, sigma_2, sigma_3]
+
+    pops = []
+    rots = [ops.X ** 0, ops.X ** 0.5, ops.Y ** 0.5]
+    for i in range(3):
+        for j in range(3):
+            circ = circuit.copy()
+            circ.append(rots[i](q_1))
+            circ.append(rots[j](q_0))
+            pops.extend(_measurement(circ))
+
+    mat = numpy.zeros((27, 15))
+    s = numpy.array([[1.0, 1.0, 1.0], [-1.0, 1.0, -1.0], [1.0, -1.0, -1.0]])
+    mat[0:3, [2, 11, 14]] = s * numpy.tile([1.0, 1.0, 1.0], (3, 1))
```

Comment:
There's way too many constants here. Is there a way to generate them automatically? At the very least they need to be documented.

Reply:
Thanks for pointing this out. Yes, it was difficult to read. The idea is to build and solve an over-determined set of linear equations of the form M*C = P, where C is a vector of length 15 containing the coefficients for the operator-sum representation of the density matrix \rho = \sum_{i, j = 0, 1, 2, 3} c_{ij} \sigma_i *\ sigma_j (excluding c_00 which is set to 1). P is a vector of length 27 containing the probabilities P_00, P_01, P_10 from the nine measurements after different basis rotations. M is a 27 x 15 matrix where each row contains three non-zero elements \in {-1, 1} which one can determine from the basis rotations. I initially just did this on paper and hard-coded the matrix M.

In the new code, M is generated more compactly within the nested for loop. Hopefully this looks less cluttered. I have also added more explanation of the above method in the doc string.